### PR TITLE
release-lib: build packages without `meta.platforms` on all platforms

### PR DESCRIPTION
**Pull request by emilazy**
This was added in <https://github.com/NixOS/nixpkgs/pull/19990> to ensure that packages that previously weren’t building at all would get at least one platform’s worth of coverage. However, since most `runCommand` calls don’t explicitly set any `meta` information, this meant that things like `tests.makeWrapper` were only being run for `x86_64-linux`. Since most trivial builder calls should work on all platforms, it doesn’t make much sense for them to be restricted in this way, and we shouldn’t unnecessarily penalize other platforms by treating an empty `meta.platforms` inconsistently with `check-meta`’s maximally‐permissive interpretation. Actual packages that only work on a subset of platforms should, of course, set `meta.platforms` explicitly.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
(For posterity: I talked myself into the more radical version of this after seeing that the `[ "x86-64-linux" ]` default was not, as I had assumed, “we’re expanding platform support but old derivations without anything set will assume they only run on the default platform” but instead “these things aren’t building on Hydra at *all* and we should fix that”.)

**Review by philiptaron — APPROVED** [review]
Heck yes.

**Review by philiptaron — APPROVED** [review]

**Review by RossComputerGuy — APPROVED** [review]
i approb

**Comment by emilazy** [comment]
Don’t get too excited, I’m still playing whack‐a‐mole with the evaluation failures. But hopefully we can get this in.

**Review by philiptaron — APPROVED** [review]
Master is now questionable, but I leave it up to the author about whether to retarget the PR to staging.

**Comment by emilazy** [comment]
The latest push should be a net addition of about ~5k Hydra jobs by setting `hydraPlatforms` for TeX Live constituent packages, after discussion in the Staging room on Matrix. Some ~2.5k of the remaining jobs are actual useful SBCL and PHP packages that aren’t getting built for `aarch64-linux` or Darwin right now, and there are other, smaller package sets that will be built on other platforms where they weren’t before. All of this is good stuff that we ought to be building on supported platforms (given our decision to `recurseIntoAttrs` them in the first place), and that probably was never intended to be restricted to `x86_64-linux`.

There will also be jobs that don’t generate new builds, as discussed above. I think it should be okay as it is now, based on the discussion, and we can set more `meta.{platforms,hydraPlatforms}` if anything turns up that looks particularly problematic. I’d prefer for someone else to merge this, after seeing ofborg results (which will inflate the count a bit by not including the *reduction* in `texlivePackages.*`).

File: pkgs/top-level/php-packages.nix
Comment (jtojnar) [inline pkgs/top-level/php-packages.nix:466]: I agree that the other instances should use `meta.broken` but this one specifically is needed to be filtered out on PHP > 8.3 since the PHP 8.4 source tree stopped shipping it and this broken derivation would shadow the fallback definition above.

Comment (emilazy) [reply pkgs/top-level/php-packages.nix:466]: Thanks for the insight. I’ve restored the logic using `lib.optionals` instead since it’s just this one package; let me know if you think that’s okay?

**Comment by emilazy** [comment]
> Are we going to want to get this in before ZHF and before the merge?

I missed this comment. Yes, I think so, assuming we agree that it seems like a good idea to merge, although if it happens with the `staging-next` merge then it can also just target `master` since it makes no difference in terms of additional jobs.

File: pkgs/top-level/php-packages.nix
Comment (jtojnar) [reply pkgs/top-level/php-packages.nix:466]: Makes sense to me.

**Comment by emilazy** [comment]
ofborg results now list 9982 jobs, so minus the ~4445 TeX Live jobs that this is removing, it should be a net increase of 5.5k jobs, as expected. Should be ready when anyone feels brave enough.

**Comment by RossComputerGuy** [comment]
> I missed this comment. Yes, I think so, assuming we agree that it seems like a good idea to merge, although if it happens with the `staging-next` merge then it can also just target `master` since it makes no difference in terms of additional jobs.

Sounds good, we probably should merge this in the next few hours or any time before 6PM PST to get the ball rolling. I hope to do the merge as listed in the 24.11 release schedule after work which would mean some time after 6PM PST.

**Comment by emilazy** [comment]
Sure, I guess it could be merged now and another `staging-next` eval kicked off. Or it could be merged directly before the merge into `master`. I don’t know really; I don’t understand Hydra bottlenecks well enough to say what’s best and it was enough of a pain to figure out what was going on here and come up with a safe‐seeming plan to address it, so I’ll let someone else do the scary merge.

**Comment by RossComputerGuy** [comment]
Guess I'll merge it.

**Comment by emilazy** [comment]
Not sure whether to kick off a `staging-next` evaluation or not. There’s an installer test issue that is currently blocking the merge into `master`.

**Comment by RossComputerGuy** [comment]
> Not sure whether to kick off a `staging-next` evaluation or not.

Idk how to kick one off but sure, why not? Better get some of that work rolling.

> There’s an installer test issue that is currently blocking the merge into `master`.

Is there an open issue? I'll look into that before merging staging-next.

**Comment by emilazy** [comment]
Well, because there’ll presumably be a new eval to test the installer fix anyway. See https://github.com/NixOS/nixpkgs/pull/348827#issuecomment-2441081863. We cannot merge as it would block the channel.

File: pkgs/top-level/php-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/php-packages.nix:651]: Isn't this wrong now? Before it was enabled on linux and now it is marked broken on linux.

Comment (emilazy) [reply pkgs/top-level/php-packages.nix:651]: Yes, oops. But it turns out it builds fine on Darwin now. https://github.com/NixOS/nixpkgs/pull/353479

Comment (SuperSandro2000) [reply pkgs/top-level/php-packages.nix:651]: I think pdo had the same mistake #353492

**Pull request merged**
